near-sdk/src/store/bucket/iter.rs

```diff
@@ -0,0 +1,210 @@
+{
+    type Item = &'a mut T;
+    type IntoIter = IterMut<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter_mut()
+    }
+}
+
+/// An iterator over elements in the storage bucket. This only yields the occupied entries.
+pub struct Iter<'a, T>
+where
+    T: BorshDeserialize + BorshSerialize,
+{
+    /// Values iterator which contains empty and filled cells.
+    values: vec::Iter<'a, Container<T>>,
+    /// Amount of valid elements left to iterate.
+    elements_left: u32,
+}
+
+impl<'a, T> Iter<'a, T>
+where
+    T: BorshDeserialize + BorshSerialize,
+{
+    pub(super) fn new(bucket: &'a Bucket<T>) -> Self {
+        Self { values: bucket.elements.iter(), elements_left: bucket.occupied_count }
+    }
+    fn decrement_elements(&mut self) {
+        self.elements_left = self
+            .elements_left
+            .checked_sub(1)
+            .unwrap_or_else(|| env::panic_str(ERR_INCONSISTENT_STATE));
+    }
+}
+
+impl<'a, T> Iterator for Iter<'a, T>
+where
+    T: BorshDeserialize + BorshSerialize,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.elements_left == 0 {
+            return None;
+        }
+        loop {
+            match self.values.next() {
+                Some(Container::Empty { .. }) => continue,
+                Some(Container::Occupied(value)) => {
+                    self.decrement_elements();
+                    return Some(value);
+                }
+                None => {
+                    // This should never be hit, because if 0 occupied elements, should have
+                    // returned before the loop
+                    env::panic_str(ERR_INCONSISTENT_STATE)
+                }
+            }
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let elements_left = self.elements_left as usize;
+        (elements_left, Some(elements_left))
+    }
+
+    fn count(self) -> usize {
+        self.elements_left as usize
+    }
+}
+
+impl<'a, T> ExactSizeIterator for Iter<'a, T> where T: BorshSerialize + BorshDeserialize {}
+impl<'a, T> FusedIterator for Iter<'a, T> where T: BorshSerialize + BorshDeserialize {}
+
+impl<'a, T> DoubleEndedIterator for Iter<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        if self.elements_left == 0 {
+            return None;
+        }
+        loop {
+            match self.values.next_back() {
+                Some(Container::Empty { .. }) => continue,
+                Some(Container::Occupied(value)) => {
+                    self.decrement_elements();
+                    return Some(value);
+                }
+                None => {
+                    // This should never be hit, because if 0 occupied elements, should have
+                    // returned before the loop
+                    env::panic_str(ERR_INCONSISTENT_STATE)
+                }
+            }
+        }
```

Comment:
So this actually can't be done cleanly because you can't make the type generic with the different bounds because they are wrapped within the `Container` type.

the duplicate `next` functions can't be deduplicated because the references have different mutability, and `next` vs `next_back` can't be duplicated because they can't be generic on what function is called.

Code can be deduplicated with a declarative macro, but it comes at the cost of readability so I'm a little hesitant to make that change.

Reply:
ahh makes sense, then definitely don't need to do this then